src/main/java/net/neoforged/neoforge/transfer/handlers/templates/AggregateResourceHandler.java

```diff
@@ -0,0 +1,194 @@
+    protected final IResourceHandler<T>[] handlers;
+
+    @SafeVarargs
+    public AggregateResourceHandler(IResourceHandler<T>... handlers) {
+        this.handlers = handlers;
+    }
+
+    @Override
+    public int size() {
+        return Stream.of(getHandlers()).mapToInt(IResourceHandler::size).sum();
+    }
+
+    @Override
+    public T getResource(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.getResource(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public int getAmount(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.getAmount(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public int getCapacity(int index, T resource) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.getCapacity(index, resource);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public int getCapacity(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.getCapacity(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public boolean isValid(int index, T resource) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.isValid(index, resource);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public boolean allowsInsertion(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.allowsInsertion(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public boolean allowsExtraction(int index) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                return storage.allowsExtraction(index);
+            }
+            index -= storage.size();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public int insert(int index, T resource, int amount, TransferAction action) {
+        for (IResourceHandler<T> storage : getHandlers()) {
+            if (index < storage.size()) {
+                if (storage.allowsInsertion()) {
```

Comment:
Do we need to check this? I feel like we should be able to just assume that the inner storage will short circuit if they aren't allowing insertion. That way if their check is for some reason not trivial, then it only has to be done once.